```markdown
# 🍽️ Multinational Restaurant - Sentiment Analysis

This notebook demonstrates a complete step-by-step pipeline for cleaning and preprocessing text reviews from a multinational restaurant dataset to prepare them for natural language processing (NLP) and machine learning classification.
```

In [ ]:
# --- Project Title: Multinational Restaurant Sentiment Analysis ---
# The goal of this project is to build a Natural Language Processing (NLP) pipeline
# to classify customer restaurant reviews as either positive (1) or negative (0).

```markdown
### 📝 File Types and Dataset Format
Understanding differences between standard file delimiters:
* **CSV** (Comma Separated Values): Columns are separated by commas (`,`).
* **TSV** (Tab Separated Values): Columns are separated by tabs (`\t`). TSV is highly preferred for text processing because reviews often contain natural commas.
```

In [ ]:
# --- Example of a CSV (Comma-Separated Values) format ---
# In CSV files, columns are separated by a comma.
# Format: name,prod,review,sentiment

In [ ]:
# --- Example of a TSV (Tab-Separated Values) format ---
# In TSV files, columns are separated by tabs (\t), which is safer for text containing natural commas.
# Format: name\tprod\treview\tsentiment

In [ ]:
# --- Data Wrangling and Exploratory Libraries ---
# We will use Pandas for data structured tables and NumPy for numerical operations.

```markdown
## 🛠️ Step 1: Import Libraries & Load Dataset
We begin by setting up our analytical environment with `pandas` and `numpy`, then reading the TSV file using safety arguments to prevent parsing issues.

In [ ]:
# --- Library Definitions ---
# Pandas: High-performance data structure (DataFrame) used for cleaning, manipulation, and analysis.
# NumPy: Fundamental library for scientific and mathematical array operations.

In [1]:
!pip install pandas

In [2]:
# --- Official Reference link for Pandas Package documentation and updates ---
# https://pypi.org/project/pandas/

In [3]:
# Import Pandas library under the standard alias 'pd' for easier syntax reference
import pandas as pd

In [4]:
# Load the tab-separated dataset safely
# - delimiter='\t' specifies that the data columns are separated by tabs
# - quoting=3 ignores double quotes to prevent parsing errors when reading raw text
df = pd.read_csv("/content/Restaurant_Reviews.tsv", delimiter="\t", quoting=3)

In [5]:
# Render and display the loaded DataFrame to inspect its records, columns, and length
df

,Review,Liked
0,Wow... Loved this place.,1
1,Crust is not good.,0
2,Not tasty and the texture was just nasty.,0
3,Stopped by during the late May bank holiday of...,1
4,The selection on the menu was great and so wer...,1
...,...,...
995,I think food should have flavor and texture an...,0
996,Appetite instantly gone.,0
997,Overall I was not impressed and would not go b...,0
998,"The whole experience was underwhelming, and I ...",0


In [6]:
# --- Terminology & Structural definitions in Pandas ---
# Dataset: The entire pool of loaded historical data.
# DataFrame: A 2-dimensional labeled tabular data structure containing multiple columns/features.
# Series: A 1-dimensional labeled array representing a single column/feature.

In [7]:
# --- Exploratory Data Analysis (EDA) Phase ---
# Visualizing, summarizing, and understanding statistical characteristics of the dataset.

```markdown
## 🔍 Step 2: Exploratory Data Analysis (EDA)
Let's analyze the dataset's characteristics, dimensions, column properties, and inspect samples of the reviews to understand patterns or inconsistencies.

In [8]:
# Display structural summary of the DataFrame (non-null counts, data types, memory usage)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Review  1000 non-null   object
 1   Liked   1000 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 15.8+ KB


In [9]:
# Check the dimensionality of the dataset in (rows, columns) format
df.shape

(1000, 2)

In [10]:
# Display the names of all headers/columns present in the DataFrame
df.columns

Index(['Review', 'Liked'], dtype='object')

In [11]:
# Display the top 5 records of the dataset for a quick overview
df.head()

,Review,Liked
0,Wow... Loved this place.,1
1,Crust is not good.,0
2,Not tasty and the texture was just nasty.,0
3,Stopped by during the late May bank holiday of...,1
4,The selection on the menu was great and so wer...,1


In [12]:
# Display the top 10 records to inspect initial reviews
df.head(10)

,Review,Liked
0,Wow... Loved this place.,1
1,Crust is not good.,0
2,Not tasty and the texture was just nasty.,0
3,Stopped by during the late May bank holiday of...,1
4,The selection on the menu was great and so wer...,1
5,Now I am getting angry and I want my damn pho.,0
6,Honeslty it didn't taste THAT fresh.),0
7,The potatoes were like rubber and you could te...,0
8,The fries were great too.,1
9,A great touch.,1


In [13]:
# Display the top 20 records to look for pattern trends and noise types
df.head(20)

,Review,Liked
0,Wow... Loved this place.,1
1,Crust is not good.,0
2,Not tasty and the texture was just nasty.,0
3,Stopped by during the late May bank holiday of...,1
4,The selection on the menu was great and so wer...,1
5,Now I am getting angry and I want my damn pho.,0
6,Honeslty it didn't taste THAT fresh.),0
7,The potatoes were like rubber and you could te...,0
8,The fries were great too.,1
9,A great touch.,1


In [14]:
# Display the top 50 records to evaluate larger samples of text distribution
df.head(50)

,Review,Liked
0,Wow... Loved this place.,1
1,Crust is not good.,0
2,Not tasty and the texture was just nasty.,0
3,Stopped by during the late May bank holiday of...,1
4,The selection on the menu was great and so wer...,1
5,Now I am getting angry and I want my damn pho.,0
6,Honeslty it didn't taste THAT fresh.),0
7,The potatoes were like rubber and you could te...,0
8,The fries were great too.,1
9,A great touch.,1


In [15]:
# Display the last 5 records of the dataset to verify trailing data pattern
df.tail()

,Review,Liked
995,I think food should have flavor and texture an...,0
996,Appetite instantly gone.,0
997,Overall I was not impressed and would not go b...,0
998,"The whole experience was underwhelming, and I ...",0
999,"Then, as if I hadn't wasted enough of my life ...",0


In [16]:
# Display the last 10 records of the dataset
df.tail(10)

,Review,Liked
990,The refried beans that came with my meal were ...,0
991,Spend your money and time some place else.,0
992,A lady at the table next to us found a live gr...,0
993,the presentation of the food was awful.,0
994,I can't tell you how disappointed I was.,0
995,I think food should have flavor and texture an...,0
996,Appetite instantly gone.,0
997,Overall I was not impressed and would not go b...,0
998,"The whole experience was underwhelming, and I ...",0
999,"Then, as if I hadn't wasted enough of my life ...",0


In [17]:
# Display the last 20 records of the dataset
df.tail(20)

,Review,Liked
980,I have been to very few places to eat that und...,0
981,We started with the tuna sashimi which was bro...,0
982,Food was below average.,0
983,It sure does beat the nachos at the movies but...,0
984,"All in all, Ha Long Bay was a bit of a flop.",0
985,The problem I have is that they charge $11.99 ...,0
986,Shrimp- When I unwrapped it (I live only 1/2 a...,0
987,"It lacked flavor, seemed undercooked, and dry.",0
988,It really is impressive that the place hasn't ...,0
989,I would avoid this place if you are staying in...,0


In [18]:
# Insight as per data exploration
  # Remove special char, punctuation , keep only alphabets
  # Text Normalization -- lowercasing (avoid redendunt word)
  # Need to remove stopwords
  # Reduce/Convert word to root/base form(avoid same meaning word from diffent form, reduce curse of dimentational problem) - stemming/lemmatization



```markdown
## 🧹 Step 3: Text Cleaning & Preprocessing Pipeline
To feed raw text into a Machine Learning model, we must standardize it. Our preprocessing pipeline covers:
1. **Regular Expression Cleaning**: Keeping only alphabetic characters.
2. **Text Normalization**: Standardizing case to lower-case.
3. **Stopwords Elimination**: Filtering out common words that don't carry sentiment value while **retaining** polar negative indicators (such as *not*, *no*, *never*).
4. **Stemming**: Reducing inflected words to their base stem root form using the `PorterStemmer`.
5. **Corpus Formulation**: Re-assembling clean words and compiling them into a final analytical collection.

In [19]:
# --- Text Preprocessing Pipeline Step-by-Step ---
# Clean, structure, and standardize raw unstructured text reviews.

In [20]:
# Simple range iteration loop testing index counter
for i in range(10):
    print(i)

0
1
2
3
4
5
6
7
8
9


In [21]:
# Loop setup to dynamically iterate over all 1000 reviews in our dataset
# for i in range(0, 1000):
#     print(i)

In [22]:
# Import Python's built-in Regular Expression library (re) for string pattern matching
import re

In [23]:
# Define sample text containing special characters, brackets, and numbers to test cleaning
text = "A regular expression (shortened as regex or regexp),[1] sometimes referred to as a rational expression,[2][3] is a sequence of characters that specifies a match pattern in text."

In [24]:
# Apply regex substitution: keep only alphabetic characters [a-zA-Z] and replace everything else with space
re.sub(pattern="[^a-zA-Z]", repl=" ", string=text)

'A regular expression  shortened as regex or regexp      sometimes referred to as a rational expression        is a sequence of characters that specifies a match pattern in text '

In [25]:
# Inspect the 'Review' series containing raw text comments
df['Review']

,Review
0,Wow... Loved this place.
1,Crust is not good.
2,Not tasty and the texture was just nasty.
3,Stopped by during the late May bank holiday of...
4,The selection on the menu was great and so wer...
...,...
995,I think food should have flavor and texture an...
996,Appetite instantly gone.
997,Overall I was not impressed and would not go b...
998,"The whole experience was underwhelming, and I ..."


In [26]:
# Select and print the first review from the dataset
df['Review'][0]

'Wow... Loved this place.'

In [27]:
# Select and print the second review
df['Review'][1]

'Crust is not good.'

In [28]:
# Select and print the third review
df['Review'][2]

'Not tasty and the texture was just nasty.'

In [29]:
# Select and print the seventh review
df['Review'][6]

"Honeslty it didn't taste THAT fresh.)"

In [30]:
# Remove all non-alphabetic characters from the seventh review
review_7 = re.sub(pattern="[^a-zA-Z]", repl=" ", string=df['Review'][6])

In [31]:
# Convert the cleaned seventh review to lowercase to normalize formatting
review_7.lower()

'honeslty it didn t taste that fresh  '

In [32]:
# View columns list to ensure alignment during cleaning loops
df.columns

Index(['Review', 'Liked'], dtype='object')

In [33]:
# View values of target variable column 'Liked' representing customer feedback category
df['Liked']

,Liked
0,1
1,0
2,0
3,1
4,1
...,...
995,0
996,0
997,0
998,0


In [34]:
# Step 1: Remove special characters and punctuation from first review
review_1 = re.sub(pattern="[^a-zA-Z]", repl=" ", string=df['Review'][0])
print("text_without special char:\n", review_1)
print("======================================")

# Step 2: Convert to lowercase
print("Normalizad Review:\n", review_1.lower())

text_without special char:
 Wow    Loved this place 
Normalizad Review:
 wow    loved this place 


In [35]:
# Install the Natural Language Toolkit (NLTK) to perform tokenization, stopwords filtration, and stemming
!pip install nltk

In [36]:
# Import the nltk library to configure our language tools
import nltk

In [37]:
# Download standard package resources containing the list of English stopwords
nltk.download("stopwords")

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [38]:
# Import the stopwords dataset from NLTK corpus package
from nltk.corpus import stopwords

In [39]:
# Fetch English stopwords dictionary dataset and save it to 'sw'
sw = stopwords.words("english")

In [40]:
# Print standard nltk English stopword array to view target words
sw

['a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 "he'd",
 "he'll",
 'her',
 'here',
 'hers',
 'herself',
 "he's",
 'him',
 'himself',
 'his',
 'how',
 'i',
 "i'd",
 'if',
 "i'll",
 "i'm",
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it'd",
 "it'll",
 "it's",
 'its',
 'itself',
 "i've",
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'on

In [41]:
# Define the list of negative indicator stopwords we want to PRESERVE
# preserving these words ensures we do not lose polarity meaning (e.g. "not good")
negative_words = {
    "not", "no", "nor", "never", "don", "don't", "ain", "aren", "aren't",
    "couldn", "couldn't", "didn", "didn't", "doesn", "doesn't", "hadn", "hadn't",
    "hasn", "hasn't", "haven", "haven't", "isn", "isn't", "ma", "mightn", "mightn't",
    "mustn", "mustn't", "needn", "needn't", "shan", "shan't", "shouldn", "shouldn't",
    "wasn", "wasn't", "weren", "weren't", "won", "won't", "wouldn", "wouldn't"
}

# Create our custom stopword list by excluding the critical negative indicators
custom_stopwords = set(sw) - negative_words

In [42]:
# Display the custom stopword set to verify that words like "not" or "no" are excluded
custom_stopwords

{'a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'd',
 'did',
 'do',
 'does',
 'doing',
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'has',
 'have',
 'having',
 'he',
 "he'd",
 "he'll",
 "he's",
 'her',
 'here',
 'hers',
 'herself',
 'him',
 'himself',
 'his',
 'how',
 'i',
 "i'd",
 "i'll",
 "i'm",
 "i've",
 'if',
 'in',
 'into',
 'is',
 'it',
 "it'd",
 "it'll",
 "it's",
 'its',
 'itself',
 'just',
 'll',
 'm',
 'me',
 'more',
 'most',
 'my',
 'myself',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'only',
 'or',
 'other',
 'our',
 'ours',
 'ourselves',
 'out',
 'over',
 'own',
 're',
 's',
 'same',
 'she',
 "she'd",
 "she'll",
 "she's",
 'should',
 "should've",
 'so',
 'some',
 'such',
 't',
 'than',
 'that',
 "that'll",
 'the',
 'their',
 'theirs',
 'them',
 'themselves',
 'then',
 't

In [43]:
# Step 1: Remove special characters and punctuation from first review
review_1 = re.sub(pattern="[^a-zA-Z]", repl=" ", string=df['Review'][0])
print("text_without special char:\n", review_1)
print("======================================")

# Step 2: Normalize cases by lowercasing
review_normalization = review_1.lower()
print("Normalizad Review:\n", review_normalization)
print("======================================")

# Step 3: Remove custom stopwords via list comprehension
print("Review without stopwords:\n")
review_without_stopwords = [word for word in review_normalization.split() if word not in custom_stopwords]
print(review_without_stopwords)

text_without special char:
 Wow    Loved this place 
Normalizad Review:
 wow    loved this place 
Review without stopwords:

['wow', 'loved', 'place']


In [44]:
# Import PorterStemmer for stemming and map initial stemmer object
from nltk.stem import PorterStemmer
ps = PorterStemmer()

In [45]:
# Step 1: Remove special characters from first review
review_1 = re.sub(pattern="[^a-zA-Z]", repl=" ", string=df['Review'][0])
print("text_without special char:\n", review_1)
print("======================================")

# Step 2: Case normalization
review_normalization = review_1.lower()
print("Normalizad Review:\n", review_normalization)
print("======================================")

# Step 3: Remove custom stopwords
print("Review without stopwords:\n")
review_without_stopwords = [word for word in review_normalization.split() if word not in custom_stopwords]
print(review_without_stopwords)
print("======================================")

# Step 4: Reduce words to root form using PorterStemmer (Stemming)
print("Review with stem word:\n")
review_with_stem_word = [ps.stem(word) for word in review_without_stopwords]
print(review_with_stem_word)

text_without special char:
 Wow    Loved this place 
Normalizad Review:
 wow    loved this place 
Review without stopwords:

['wow', 'loved', 'place']
Review with stem word:

['wow', 'love', 'place']


In [46]:
# Verify content of the original first review for reference comparisons
df['Review'][0]

'Wow... Loved this place.'

In [47]:
# Display comparison between the raw original review and the final cleaned representation
actual_review = "Wow... Loved this place."
print("actual review:\n", actual_review)
print("======================================")
clean_review = ['wow', 'love', 'place']
print("actual review with cleaning:\n", " ".join(clean_review))

actual review:
 Wow... Loved this place.
actual review with cleaning:
 wow love place


In [48]:
corpus = []

# # Step1:Remove special char, punctuation , keep only alphabets
review_1 = re.sub(pattern="[^a-zA-Z]",repl=" ",string=df['Review'][0])
print("step1: text_without special char:\n",review_1)
print("======================================")

# Step2:Text Normalization -- lowercasing (avoid redendunt word)
review_normalization =  review_1.lower()
print("Step2: Normalizad Review:\n",review_normalization)

# Step3:Need to remove stopwords
print("======================================")
print("Step3 : Review without stopwords:\n")
# for word in review_normalization.split():
#   if word not in sw:
#     print(word)


# review_without_stopwords = []
# for word in review_normalization.split(): # for loop
#   if word not in sw:# if cond
#     review_without_stopwords.append(word)  # expression

# print(review_without_stopwords)


# list compression
# [expression for loop if condition]
review_without_stopwords= [ word for word in review_normalization.split()  if word not in custom_stopwords ]

print(review_without_stopwords)

# Step4:# Reduce/Convert word to root/base form
print("======================================")
print("Step4: Review with stem word:\n")
# for word in review_without_stopwords:
#   print(ps.stem(word))


# list compression
# [expression for loop]

review_with_stem_word = [ps.stem(word) for word in review_without_stopwords]
print(review_with_stem_word)

# Step5:Joining the stemmed words
review =" ".join(review_with_stem_word)

print("======================================")
print("Step5: Review with stem word:\n",review)

print("======================================")
# step 6: append word to corpus
corpus.append(review)
print("Step 6: review stored in corpus")
print(corpus)

step1: text_without special char:
 Wow    Loved this place 
Step2: Normalizad Review:
 wow    loved this place 
Step3 : Review without stopwords:

['wow', 'loved', 'place']
Step4: Review with stem word:

['wow', 'love', 'place']
Step5: Review with stem word:
 wow love place
Step 6: review stored in corpus
['wow love place']


In [49]:
# Initialize clean corpus collection list
corpus = []

# Loop over all 1000 dataset reviews and process systematically
for i in range(0, 1000):
    # Step 1: Clean non-alphabetic characters
    review_without_special_char = re.sub(pattern="[^a-zA-Z]", repl=" ", string=df['Review'][i])

    # Step 2: Standardize letter casing
    review_normalization = review_without_special_char.lower()

    # Step 3: Remove custom stopwords
    review_without_stopwords = [word for word in review_normalization.split() if word not in custom_stopwords]

    # Step 4: Reduce terms to root structure
    ps = PorterStemmer()
    review_with_stem_word = [ps.stem(word) for word in review_without_stopwords]

    # Step 5: Compile list into processed text structure
    review = " ".join(review_with_stem_word)

    # Step 6: Save processed text to corpus
    corpus.append(review)

In [50]:
# Validate the raw first review text contents
df['Review'][0]

'Wow... Loved this place.'

In [51]:
# Validate the raw second review text contents
df['Review'][1]

'Crust is not good.'

In [52]:
# Validate the raw third review text contents
df['Review'][2]

'Not tasty and the texture was just nasty.'

In [53]:
# Print the finished corpus list displaying clean text segments ready for modeling
print(corpus)

['wow love place', 'crust not good', 'not tasti textur nasti', 'stop late may bank holiday rick steve recommend love', 'select menu great price', 'get angri want damn pho', 'honeslti didn tast fresh', 'potato like rubber could tell made ahead time kept warmer', 'fri great', 'great touch', 'servic prompt', 'would not go back', 'cashier no care ever say still end wayyy overpr', 'tri cape cod ravoli chicken cranberri mmmm', 'disgust pretti sure human hair', 'shock no sign indic cash', 'highli recommend', 'waitress littl slow servic', 'place not worth time let alon vega', 'not like', 'burritto blah', 'food amaz', 'servic also cute', 'could care less interior beauti', 'perform', 'right red velvet cake ohhh stuff good', 'never brought salad ask', 'hole wall great mexican street taco friendli staff', 'took hour get food tabl restaur food luke warm sever run around like total overwhelm', 'worst salmon sashimi', 'also combo like burger fri beer decent deal', 'like final blow', 'found place acci

In [54]:
# --- Insights and Summary of Text Preprocessing ---

# This section focuses on preparing the raw text reviews for machine learning by applying a series of preprocessing techniques.
# The goal is to clean the text, standardize it, and reduce its dimensionality, which are crucial steps for effective text analysis.

# 1. Removal of Special Characters and Punctuation:
#    - Using regular expressions (`re.sub(pattern="[^a-zA-Z]", repl=" ", string=...)`), all characters that are not English alphabet letters
#      are replaced with spaces. This helps to remove noise such as numbers, symbols, and extra punctuation that do not contribute to sentiment analysis.

# 2. Text Normalization (Lowercasing):
#    - All words are converted to lowercase (`.lower()`). This ensures that words like 'Wow', 'wow', and 'WOW' are treated as the same word,
#      preventing the model from learning redundant features and improving consistency.

# 3. Stopword Removal:
#    - Common words (stopwords like 'the', 'a', 'is') that carry little to no semantic meaning for sentiment analysis are removed.
#    - A custom stopword list (`custom_stopwords`) was created by modifying the NLTK English stopwords to *retain* negative words
#      (e.g., 'not', 'no', 'don't'). This is a crucial step to preserve the sentiment polarity of reviews, as these words directly impact meaning.
#      For example, 'food was good' and 'food was not good' have opposite sentiments, and removing 'not' would lose this distinction.

# 4. Stemming:
#    - Using NLTK's `PorterStemmer`, words are reduced to their root form (e.g., 'loved' becomes 'love', 'eating' becomes 'eat').
#    - This technique (or lemmatization, a more linguistically informed approach) helps in reducing the vocabulary size and treating
#      different inflections of a word as a single term, further reducing dimensionality and improving model generalization.

# 5. Corpus Creation:
#    - After applying all these steps, the cleaned and processed words for each review are joined back into a string.
#    - These processed review strings are then collected into a list called `corpus`. This `corpus` now represents the clean, preprocessed
#      text data, ready for the next stage: feature extraction (e.g., Bag-of-Words or TF-IDF) to convert text into numerical format for machine learning models.
#    - The `corpus` effectively transforms 1000 raw reviews into a structured format suitable for building a sentiment classifier.

In [57]:
df['Review'][0:5]

,Review
0,Wow... Loved this place.
1,Crust is not good.
2,Not tasty and the texture was just nasty.
3,Stopped by during the late May bank holiday of...
4,The selection on the menu was great and so wer...


In [55]:
corpus

['wow love place',
 'crust not good',
 'not tasti textur nasti',
 'stop late may bank holiday rick steve recommend love',
 'select menu great price',
 'get angri want damn pho',
 'honeslti didn tast fresh',
 'potato like rubber could tell made ahead time kept warmer',
 'fri great',
 'great touch',
 'servic prompt',
 'would not go back',
 'cashier no care ever say still end wayyy overpr',
 'tri cape cod ravoli chicken cranberri mmmm',
 'disgust pretti sure human hair',
 'shock no sign indic cash',
 'highli recommend',
 'waitress littl slow servic',
 'place not worth time let alon vega',
 'not like',
 'burritto blah',
 'food amaz',
 'servic also cute',
 'could care less interior beauti',
 'perform',
 'right red velvet cake ohhh stuff good',
 'never brought salad ask',
 'hole wall great mexican street taco friendli staff',
 'took hour get food tabl restaur food luke warm sever run around like total overwhelm',
 'worst salmon sashimi',
 'also combo like burger fri beer decent deal',
 'like

In [ ]:
# Need to feed data to machine to make prediction?

In [ ]:
# Machine don't know anything

In [ ]:
# Need to feed data to machine to learn   then make prediction

In [ ]:
# 1000 reviews are cleaned (including all above step) stored inside---> corpus list

In [ ]:
# Machine cannot directly read text data

In [ ]:
# Convert Data into Numerical Forms

# Text Representation:
  #1.One hot encoding
  #2.BOW/Countvectorizer
  #3.SkipGram
  #4.TF-IDF

In [59]:
import nltk

In [74]:
# dir(nltk)

In [62]:
# help(nltk)

In [63]:
!pip install scikit-learn

In [64]:
import sklearn

In [65]:
dir(sklearn)

['calibration',
 'clone',
 'cluster',
 'compose',
 'config_context',
 'covariance',
 'cross_decomposition',
 'datasets',
 'decomposition',
 'discriminant_analysis',
 'dummy',
 'ensemble',
 'exceptions',
 'experimental',
 'externals',
 'feature_extraction',
 'feature_selection',
 'frozen',
 'gaussian_process',
 'get_config',
 'impute',
 'inspection',
 'isotonic',
 'kernel_approximation',
 'kernel_ridge',
 'linear_model',
 'manifold',
 'metrics',
 'mixture',
 'model_selection',
 'multiclass',
 'multioutput',
 'naive_bayes',
 'neighbors',
 'neural_network',
 'pipeline',
 'preprocessing',
 'random_projection',
 'semi_supervised',
 'set_config',
 'show_versions',
 'svm',
 'tree']

In [ ]:
# sklearn.feature_extraction.text - Utilities to build feature vectors from text documents

In [ ]:
# max_features : int, default=None
#      |      If not None, build a vocabulary that only consider the top
#      |      `max_features` ordered by term frequency across the corpus.
#      |      Otherwise, all features are used.
#      |
#      |      This parameter is ignored if vocabulary is not None.

In [69]:
# help(sklearn.feature_extraction.text)

In [70]:
from sklearn.feature_extraction.text import CountVectorizer

In [71]:
cv = CountVectorizer()

In [ ]:
# corpus ---> all word
# vocab  ---> all word from corpus


In [1]:
from sklearn.feature_extraction.text import CountVectorizer

# --- Text Feature Extraction using CountVectorizer ---
# CountVectorizer is a tool from scikit-learn that converts a collection of text documents
# into a matrix of token counts. It's a fundamental step in preparing text data for
# machine learning models, as models require numerical input.

# How it works:
# 1. Tokenization: Splits text into individual words (tokens).
# 2. Vocabulary Creation: Builds a dictionary of all unique words in the corpus.
# 3. Word Counting: Transforms text into a matrix where each row represents a document
#    and each column represents a word from the vocabulary, with cell values indicating
#    the frequency of that word in the document.

# The `max_features` parameter (used in the subsequent cell) is crucial for controlling
# the size of the vocabulary. If set, it instructs CountVectorizer to only consider the
# top 'n' words ordered by their frequency across the entire corpus. This helps in:
# - Reducing dimensionality: Limits the number of features (words), making the model
#   computationally less expensive and reducing the risk of overfitting.
# - Focusing on most relevant words: Often, the most frequent words are the most
#   informative for classification tasks.

In [75]:
cv = CountVectorizer(max_features=1500)

In [ ]:
# corpus ---> all word
# vocab  ---> select only 1500 unique word from corpus based upon there frequency

In [73]:
corpus[0:5]

['wow love place',
 'crust not good',
 'not tasti textur nasti',
 'stop late may bank holiday rick steve recommend love',
 'select menu great price']

The `corpus` list, displayed above, contains the first five preprocessed reviews. Each review is now a string of stemmed, lowercase words, with stopwords (excluding negative indicators) and special characters removed. This is the numerical input that `CountVectorizer` will process.

In [ ]:
# corpus = ['wow love place',
#  'crust not good',
#  'not tasti textur nasti',
#  'stop late may bank holiday rick steve recommend love',
#  'select menu great price']

This cell conceptually visualizes the word frequency count. In essence, `CountVectorizer` builds an internal vocabulary, assigning an index to each unique word. The `max_features` parameter then filters this vocabulary, retaining only the `N` most frequent words, which are often the most impactful for sentiment analysis. Words like 'wow', 'love', and 'place' appear frequently, indicating their importance, while less frequent words like 'menu', 'late', and 'rick' would be discarded if `max_features` is set to a value that excludes them.

The comments in the cell above illustrate how `CountVectorizer` prioritizes words. When `max_features` is set (e.g., to 1500 as in `cv = CountVectorizer(max_features=1500)`), the `CountVectorizer` examines all unique words across the entire `corpus`.

It then ranks these words by their frequency of appearance. The `max_features` parameter ensures that only the `N` most frequent words are kept in the vocabulary. Words with lower frequencies (like 'menu', 'late', 'rick' in the example) might be discarded if `N` is exceeded. This process reduces the dimensionality of the feature space while retaining the most informative words for the model.

In [ ]:
# wow ----> 100
# love ---> 90
# place --> 80
# crust --> 70
# ---
# ---
# ---
# ---
# ---
# menu --> 5
# late --> 2
# rick --> 1

In [76]:
cv

CountVectorizer(max_features=1500)

Here, the `CountVectorizer` object `cv` is initialized with `max_features=1500`. This is a critical step for dimensionality reduction and model efficiency. It tells the vectorizer to consider only the 1500 most frequently occurring words in our entire preprocessed text corpus when creating the Bag-of-Words model. This helps to filter out rare words that might not contribute much to sentiment and can reduce computational complexity and potential overfitting.

In [77]:
cv.fit_transform(corpus).toarray()    # 1500 -(3+3) = 1494

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [ ]:
# fit   ------> learn (learn word, pattern,learn exam --> corpus, orderby based on freq, select max feature as vocab,)
# transform --> make transformation based on learning   --> good ---> [1,0,0,0,000000000000000000000000000000000000] --> 1500
                                                      # --> love ---> [0,1,0000000000000000000000000000000000000000]

In [78]:
# Indepedant feature

X = cv.fit_transform(corpus).toarray()    # 1500 -(3+3) = 1494
X

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

This is where the `CountVectorizer` truly transforms our text data into numerical features.

*   **`cv.fit_transform(corpus)`**: This method performs two main actions:
    1.  **`fit`**: It learns the vocabulary from the `corpus` (our list of cleaned reviews). Based on the `max_features=1500` setting, it identifies the 1500 most frequent words and creates a mapping (vocabulary) from these words to unique integer indices.
    2.  **`transform`**: It then converts each review in the `corpus` into a numerical vector. Each vector will have 1500 elements (corresponding to the 1500 selected words). The value at each position indicates the count of that specific word in the review.

*   **`.toarray()`**: The output of `fit_transform` is a sparse matrix (efficient for memory when most values are zero). `.toarray()` converts this sparse matrix into a dense NumPy array, which is a standard format for many machine learning algorithms. The resulting array `X` is our independent feature set, where each row is a review, and each column represents a word count.

In [ ]:
# Depedant feature

In [80]:
df.iloc[:]

,Review,Liked
0,Wow... Loved this place.,1
1,Crust is not good.,0
2,Not tasty and the texture was just nasty.,0
3,Stopped by during the late May bank holiday of...,1
4,The selection on the menu was great and so wer...,1
...,...,...
995,I think food should have flavor and texture an...,0
996,Appetite instantly gone.,0
997,Overall I was not impressed and would not go b...,0
998,"The whole experience was underwhelming, and I ...",0


In [81]:
df.iloc[:,:]   # [row index selection,column index selection]

,Review,Liked
0,Wow... Loved this place.,1
1,Crust is not good.,0
2,Not tasty and the texture was just nasty.,0
3,Stopped by during the late May bank holiday of...,1
4,The selection on the menu was great and so wer...,1
...,...,...
995,I think food should have flavor and texture an...,0
996,Appetite instantly gone.,0
997,Overall I was not impressed and would not go b...,0
998,"The whole experience was underwhelming, and I ...",0


In [82]:
df.iloc[0:,:]   # [row index selection,column index selection]

,Review,Liked
0,Wow... Loved this place.,1
1,Crust is not good.,0
2,Not tasty and the texture was just nasty.,0
3,Stopped by during the late May bank holiday of...,1
4,The selection on the menu was great and so wer...,1
...,...,...
995,I think food should have flavor and texture an...,0
996,Appetite instantly gone.,0
997,Overall I was not impressed and would not go b...,0
998,"The whole experience was underwhelming, and I ...",0


In [83]:
df.iloc[:,0]   # [row index selection,column index selection]

,Review
0,Wow... Loved this place.
1,Crust is not good.
2,Not tasty and the texture was just nasty.
3,Stopped by during the late May bank holiday of...
4,The selection on the menu was great and so wer...
...,...
995,I think food should have flavor and texture an...
996,Appetite instantly gone.
997,Overall I was not impressed and would not go b...
998,"The whole experience was underwhelming, and I ..."


In [84]:
df.iloc[:,1]   # [row index selection,column index selection]

,Liked
0,1
1,0
2,0
3,1
4,1
...,...
995,0
996,0
997,0
998,0


In [85]:
df.iloc[:,1].values   # [row index selection,column index selection]

array([1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1,
       0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1,
       1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1,
       0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1,
       0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1,

In [87]:
# Dependant Feature
y = df.iloc[:,1].values   # [row index selection,column index selection]

Here, we are extracting the dependent variable (or target variable), `y`, from our original DataFrame. The `Liked` column (`df.iloc[:, 1]`) contains the sentiment labels (1 for positive, 0 for negative). We convert these labels to a NumPy array using `.values` to align with the format required for scikit-learn models. This `y` array will be used to train our model to predict sentiment.

In [88]:
y

array([1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1,
       0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1,
       1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1,
       1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1,
       0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1,
       0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1,

In [89]:
X

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [90]:
from sklearn.model_selection import train_test_split

In [91]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.20,random_state=0)
# X_train,X_test,y_train,y_test = train_test_split(X,y,train_size=0.80,random_state=0)

In [93]:
X_train.shape,y_train.shape

((800, 1500), (800,))

In [94]:
X_test.shape,y_test.shape

((200, 1500), (200,))

In [95]:
X_train

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [96]:
y_train

array([1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1,
       0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1,
       0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1,
       1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0,
       1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0,
       0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1,
       0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1,
       1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1,

In [97]:
# testing
X_test

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [98]:
# testing
y_test

array([0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1,
       1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1,
       0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1,
       0, 1])

In this step, we split our dataset into training and testing sets using `train_test_split` from `sklearn.model_selection`.

*   **`X`**: Our independent features (the word count vectors).
*   **`y`**: Our dependent variable (the sentiment labels).
*   **`test_size=0.20`**: Specifies that 20% of the data will be used for the testing set, and the remaining 80% for the training set.
*   **`random_state=0`**: Ensures reproducibility. If you run this code again, you'll get the same split every time.

This split creates four new datasets:
*   **`X_train`**: Features for training the model (800 reviews).
*   **`X_test`**: Features for evaluating the model (200 reviews).
*   **`y_train`**: Labels corresponding to `X_train`.
*   **`y_test`**: Labels corresponding to `X_test`.

In [ ]:
# training/build machine learning model

In [99]:
from sklearn.naive_bayes import GaussianNB

In [100]:
classifier = GaussianNB()

In [101]:
classifier.fit(X_train,y_train)

GaussianNB()

Here, we are preparing and training our classification model:

1.  **`from sklearn.naive_bayes import GaussianNB`**: Imports the Gaussian Naive Bayes classifier from scikit-learn. Gaussian Naive Bayes is a probabilistic classifier based on Bayes' theorem, assuming that features follow a Gaussian (normal) distribution. It's often a good baseline for text classification tasks due to its simplicity and efficiency.

2.  **`classifier = GaussianNB()`**: Initializes an instance of the Gaussian Naive Bayes classifier.

3.  **`classifier.fit(X_train, y_train)`**: This is the training step. The classifier learns the patterns and relationships between the word count features (`X_train`) and their corresponding sentiment labels (`y_train`). During this process, the model estimates the probabilities for each word given a positive or negative sentiment, forming its internal knowledge base.

In [ ]:
# Make prediction on testing set(X_test)

In [102]:
classifier.predict(X_test)

array([1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0,
       0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1,
       1, 1])

In [103]:
y_pred = classifier.predict(X_test)

In [104]:
y_pred      # predicted sentiment of unseen review

array([1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0,
       0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1,
       1, 1])

In [105]:
y_test    # actual sentiment of unseen review

array([0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1,
       1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1,
       0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1,
       0, 1])

After training, we use the `classifier.predict(X_test)` method to make predictions on the unseen test data (`X_test`). This generates an array (`y_pred`) containing the predicted sentiment (0 or 1) for each review in the test set. We then compare these `y_pred` values with the actual `y_test` values to evaluate the model's performance.

In [106]:
from sklearn.metrics import accuracy_score

In [107]:
accuracy_score(y_test,y_pred)

0.735

In [108]:
from sklearn.metrics import confusion_matrix

In [109]:
confusion_matrix(y_test,y_pred)

array([[56, 41],
       [12, 91]])

In [111]:
from sklearn.metrics import precision_score,recall_score,f1_score

In [112]:
precision_score(y_test,y_pred)

0.6893939393939394

In [113]:
recall_score(y_test,y_pred)

0.883495145631068

In [114]:
f1_score(y_test,y_pred)

0.774468085106383

This section evaluates the performance of our trained Gaussian Naive Bayes model using various metrics:

*   **`accuracy_score(y_test, y_pred)`**: Measures the proportion of correctly classified reviews out of the total number of reviews in the test set. An accuracy of 0.735 means the model correctly predicted the sentiment for 73.5% of the reviews.

*   **`confusion_matrix(y_test, y_pred)`**: Provides a detailed breakdown of correct and incorrect predictions. It's a table showing:
    *   True Positives (top-left): Reviews correctly predicted as positive.
    *   False Positives (top-right): Reviews incorrectly predicted as positive (Type I error).
    *   False Negatives (bottom-left): Reviews incorrectly predicted as negative (Type II error).
    *   True Negatives (bottom-right): Reviews correctly predicted as negative.

*   **`precision_score(y_test, y_pred)`**: Measures the accuracy of positive predictions. It answers: "Of all reviews predicted as positive, how many were actually positive?" A precision of 0.689 indicates that about 68.9% of the reviews the model classified as positive were indeed positive.

*   **`recall_score(y_test, y_pred)`**: Measures the ability of the model to find all the positive samples. It answers: "Of all the actual positive reviews, how many did the model correctly identify?" A recall of 0.883 suggests the model captured about 88.3% of all actual positive reviews.

*   **`f1_score(y_test, y_pred)`**: The F1-score is the harmonic mean of precision and recall. It provides a single score that balances both metrics, especially useful when there's an uneven class distribution. An F1-score of 0.774 represents a good balance between precision and recall for this model.